infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,69 +141,130 @@ public LogUnitServer(ServerContext serverContext) {
      */
     @ServerHandler(type = CorfuMsgType.TAIL_REQUEST)
     public void handleTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        TailsResponse tails = batchWriter.queryTails(msg.getEpoch());
-        r.sendResponse(ctx, msg, CorfuMsgType.TAIL_RESPONSE.payloadMsg(tails));
+        log.debug("handleTailRequest: received a tail request {}", msg);
+        CorfuPayloadMsg<Void> tailsReqMsg = new CorfuPayloadMsg<>();
+        tailsReqMsg.copyBaseFields(msg);
+        batchWriter.<TailsResponse>addTask(TAILS_QUERY, tailsReqMsg)
+                .thenAccept(tailsResp -> r.sendResponse(ctx, msg, CorfuMsgType.TAIL_RESPONSE.payloadMsg(tailsResp)))
+                .exceptionally(ex -> {
+                    handleException(ex, ctx, tailsReqMsg, r);
+                    return null;
```

Comment:
I don't like that we have to return null. Can you please replace `exceptionally()` by `handle(...)` or `whenCompleted(...)`